### PR TITLE
propagate IO context more; use 2-argument `summary`

### PR DESCRIPTION
**Pull request by JeffBezanson**
Some subtle printing fixes found during #29466. Should be non-breaking.

- Set the `:module` IOContext property to Main during REPL display. I really thought we were doing this already.
- Use and define `summary(io, x)` instead of just `summary(x)` everywhere.
- Replace `print(io, "$T")` with `print(io, T)` to allow printing types to see the context in `io`. Could probably be done in more places but this catches most in Base.

### CODE REVIEW

File: base/errorshow.jl [hunk 1]
```diff
@@ -54,19 +54,19 @@ end
...
-            tstr = string(typeof(ex.got))
+            targs = (typeof(ex.got),)
         end
         if isempty(ex.context)
-            ctx = "in $(ex.func)"
+            ctx = ("in ", ex.func)
         else
-            ctx = "in $(ex.func), in $(ex.context)"
+            ctx = ("in ", ex.func, ", in ", ex.context)
         end
```
Comment (vtjnash) [inline base/errorshow.jl]: `.func` and `.context` fields are strings, they shouldn't have any special printing

File: stdlib/LinearAlgebra/src/bidiag.jl [hunk 2]
```diff
@@ -208,7 +208,7 @@ end
 
 function show(io::IO, M::Bidiagonal)
     # TODO: make this readable and one-line
-    println(io, summary(M), ":")
+    summary(io, M); println(io, ":")
```
Comment (vtjnash) [inline stdlib/LinearAlgebra/src/bidiag.jl]: doesn't seem like these need to share a line

File: stdlib/LinearAlgebra/src/bunchkaufman.jl [hunk 3]
```diff
@@ -250,7 +250,7 @@ issuccess(B::BunchKaufman) = B.info == 0
 
 function Base.show(io::IO, mime::MIME{Symbol("text/plain")}, B::BunchKaufman)
     if issuccess(B)
-        println(io, summary(B))
+        summary(io, B); println(io)
```
Comment (vtjnash) [inline stdlib/LinearAlgebra/src/bunchkaufman.jl]: doesn't seem like these need to share a line

File: stdlib/LinearAlgebra/src/cholesky.jl [hunk 4]
```diff
@@ -379,7 +379,7 @@ issuccess(C::Cholesky) = C.info == 0
 
 function show(io::IO, mime::MIME{Symbol("text/plain")}, C::Cholesky{<:Any,<:AbstractMatrix})
     if issuccess(C)
-        println(io, summary(C))
+        summary(io, C); println(io)
```
Comment (vtjnash) [inline stdlib/LinearAlgebra/src/cholesky.jl]: doesn't seem like these need to share a line

File: stdlib/LinearAlgebra/src/cholesky.jl [hunk 5]
```diff
@@ -388,7 +388,7 @@ function show(io::IO, mime::MIME{Symbol("text/plain")}, C::Cholesky{<:Any,<:Abst
 end
 
 function show(io::IO, mime::MIME{Symbol("text/plain")}, C::CholeskyPivoted{<:Any,<:AbstractMatrix})
-    println(io, summary(C))
+    summary(io, C); println(io)
```
Comment (vtjnash) [inline stdlib/LinearAlgebra/src/cholesky.jl]: doesn't seem like these need to share a line

File: stdlib/LinearAlgebra/src/eigen.jl [hunk 6]
```diff
@@ -489,7 +489,7 @@ julia> eigvecs(A, B)
 eigvecs(A::AbstractMatrix, B::AbstractMatrix) = eigvecs(eigen(A, B))
 
 function show(io::IO, mime::MIME{Symbol("text/plain")}, F::Union{Eigen,GeneralizedEigen})
-    println(io, summary(F))
+    summary(io, F); println(io)
```
Comment (vtjnash) [inline stdlib/LinearAlgebra/src/eigen.jl]: doesn't seem like these need to share a line

File: stdlib/LinearAlgebra/src/lu.jl [hunk 7]
```diff
@@ -305,7 +305,7 @@ issuccess(F::LU) = F.info == 0
 
 function show(io::IO, mime::MIME{Symbol("text/plain")}, F::LU)
     if issuccess(F)
-        println(io, summary(F))
+        summary(io, F); println(io)
```
Comment (vtjnash) [inline stdlib/LinearAlgebra/src/lu.jl]: doesn't seem like these need to share a line

File: stdlib/LinearAlgebra/src/qr.jl [hunk 8]
```diff
@@ -395,7 +395,7 @@ Matrix(F::QRPivoted) = Array(AbstractArray(F))
 Array(F::QRPivoted) = Matrix(F)
 
 function show(io::IO, mime::MIME{Symbol("text/plain")}, F::Union{QR, QRCompactWY, QRPivoted})
-    println(io, summary(F))
+    summary(io, F); println(io)
```
Comment (vtjnash) [inline stdlib/LinearAlgebra/src/qr.jl]: doesn't seem like these need to share a line

File: stdlib/LinearAlgebra/src/schur.jl [hunk 9]
```diff
@@ -116,7 +116,7 @@ Base.propertynames(F::Schur) =
     (:Schur, :vectors, fieldnames(typeof(F))...)
 
 function show(io::IO, mime::MIME{Symbol("text/plain")}, F::Schur)
-    println(io, summary(F))
+    summary(io, F); println(io)
```
Comment (vtjnash) [inline stdlib/LinearAlgebra/src/schur.jl]: doesn't seem like these need to share a line

File: stdlib/LinearAlgebra/src/schur.jl [hunk 10]
```diff
@@ -261,7 +261,7 @@ Base.propertynames(F::GeneralizedSchur) =
     (:values, :left, :right, fieldnames(typeof(F))...)
 
 function show(io::IO, mime::MIME{Symbol("text/plain")}, F::GeneralizedSchur)
-    println(io, summary(F))
+    summary(io, F); println(io)
```
Comment (vtjnash) [inline stdlib/LinearAlgebra/src/schur.jl]: doesn't seem like these need to share a line

File: stdlib/SharedArrays/src/SharedArrays.jl [hunk 11]
```diff
@@ -487,7 +487,7 @@ function show(io::IO, mime::MIME"text/plain", S::SharedArray)
         invoke(show, Tuple{IO,MIME"text/plain",DenseArray}, io, MIME"text/plain"(), S)
     else
         # retrieve from the first worker mapping the array.
-        println(io, summary(S), ":")
+        summary(io, S); println(io, ":")
```
Comment (vtjnash) [inline stdlib/SharedArrays/src/SharedArrays.jl]: doesn't seem like these need to share a line

File: stdlib/SuiteSparse/src/spqr.jl [hunk 12]
```diff
@@ -321,7 +321,7 @@ function Base.propertynames(F::QRSparse, private::Bool=false)
 end
 
 function Base.show(io::IO, mime::MIME{Symbol("text/plain")}, F::QRSparse)
-    println(io, summary(F))
+    summary(io, F); println(io)
```
Comment (vtjnash) [inline stdlib/SuiteSparse/src/spqr.jl]: doesn't seem like these need to share a line

**Comment by vtjnash** [comment]
> Replace print(io, "$T") with print(io, T) 

This makes it sound a bit like you can't stringify values before printing them. While the alternate form given here _is_ generally more efficient and preferable, it's also possible and valid to pass the `io` as a `context=` kwarg to `sprint` or `repr` AND it also valid not to (albeit losing styling). No changes necessary to this PR, but just wanted to leave this comment as a breadcrumb for anyone else who is seeing this commit and PR.

File: stdlib/LinearAlgebra/src/bidiag.jl (hunk 2 above)
Comment (JeffBezanson) [reply stdlib/LinearAlgebra/src/bidiag.jl]: Right, they don't *need* to. But nevertheless they do.

**Comment by nalimilan** [comment]
Seeing how inconvenient is the new code in many places compared with the old one which used interpolation, I think it's clear people will tend to use interpolation in their code. So I wonder whether https://github.com/JuliaLang/julia/pull/29466 is really feasible. I'm not sure we should apply the third bullet point of the description (i.e. stop using interpolation for types) if we're unsure we want to do https://github.com/JuliaLang/julia/pull/29466, given that it makes code more complex without any benefit for now.

**Comment by JeffBezanson** [comment]
I should clarify that this really isn't too important --- it only affects whether type names are printed fully qualified. That doesn't make it wrong to use interpolation. But for Base purposes like printing method errors this change makes the code more correct.

Whether this makes code more complex is perhaps subjective. While string interpolation is concise and *easy*, I would argue it is *more* complex than just passing arguments. It introduces an additional call and io context that are not apparent in the code. That is complexity.

Here's another idea: we could make `print`/`string` never include module prefixes. They are arguably "syntax-y" and hence more suited to `repr`. Then `print` on a type would always do the same thing, independent of `:module` or whether you use interpolation.

One example of a problem solved by #29466 is that you can't always develop a test case for printed output in the repl and then copy it to a test file. During testing, `using` will not happen in `Main` so you get different output. Or you might also get different output if another test injects something into `Main`, or if a user loaded some packages before interactively loading a test file.

**Comment by nalimilan** [comment]
> Here's another idea: we could make `print`/`string` never include module prefixes. They are arguably "syntax-y" and hence more suited to `repr`. Then `print` on a type would always do the same thing, independent of `:module` or whether you use interpolation.

Yeah, why not.

**Pull request merged**